frame/contracts/src/exec.rs

```diff
@@ -347,25 +370,27 @@ where
 					nested,
 				)?;
 
-				let executable = nested.loader.load_init(&code_hash)
-					.map_err(|_| Error::<T>::CodeNotFound)?;
-				let output = nested.vm
-					.execute(
-						&executable,
-						nested.new_call_context(caller.clone(), endowment),
-						input_data,
-						gas_meter,
-					).map_err(|e| ExecError { error: e.error, origin: ErrorOrigin::Callee })?;
+				let output = executable.execute(
+					nested.new_call_context(caller.clone(), endowment),
+					&ExportedFunction::Constructor,
+					input_data,
+					gas_meter,
+				).map_err(|e| ExecError { error: e.error, origin: ErrorOrigin::Callee })?;
 
 
 				// Collect the rent for the first block to prevent the creation of very large
 				// contracts that never intended to pay for even one block.
 				// This also makes sure that it is above the subsistence threshold
 				// in order to keep up the guarantuee that we always leave a tombstone behind
 				// with the exception of a contract that called `seal_terminate`.
-				Rent::<T>::charge(&dest)?
-					.and_then(|c| c.get_alive())
-					.ok_or_else(|| Error::<T>::NewContractNotFunded)?;
+				<ContractInfoOf<T>>::get(&dest)
+					.and_then(|contract| contract.get_alive())
+					.map(|contract|
+						Rent::<T, E>::charge(&dest, contract, executable.occupied_storage())
```

Comment:
My personal preference is not to have "side effects" in a `map` combinator like this. I think it would be clearer with the `if let Some(contract)` pattern.

Reply:
I would require duplicating the `Error` in the `else` branch, though. You think that is worth it?

Reply:
Maybe something like 

```
let contract = <ContractInfoOf<T>>::get(&dest)
	.and_then(|contract| contract.get_alive())
	.ok_or(Error::<T>::NewContractNotFunded)?;

Rent::<T, E>::charge(&dest, contract, executable.occupied_storage())?
	.ok_or(Error::<T>::NewContractNotFunded)?;
```

which to me communicates better what is going on, at the cost of having the error twice. I would trade that for losing `transpose()` and `flatten()`

Reply:
Also maybe they should even be different errors?

Reply:
Fair enough. I also took the chance to use two different error messages. That way we can differentiate between a contract terminating in its constructor and a contract having too little endowment (or sending it away in its constructor).